salesforce/bulkwrite.go

```diff
@@ -246,35 +275,156 @@ func (c *Connector) GetJobInfo(ctx context.Context, jobId string) (*GetJobInfoRe
+	jobInfo, err := c.GetJobInfo(ctx, jobId)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get job information: %w", err)
 	}
 
-	return c.getCSV(ctx, location)
+	if jobInfo.State != JobStateComplete {
+		// Take care of failed, aborted, in progress, and upload complete cases
+		// We don't need to query SF for these cases
+		return c.getUnsuccessfulJobResults(jobInfo), err
+	}
+
+	if jobInfo.State == JobStateComplete && jobInfo.NumberRecordsFailed == 0 {
+		// Complete success case, no need to query SF
+		return &JobResults{
+			JobId:   jobInfo.Id,
+			State:   jobInfo.State,
+			JobInfo: jobInfo,
+		}, nil
+	}
+
+	return c.getPartialFailureDetails(ctx, jobInfo)
 }
 
-func (c *Connector) GetUnprocessedResults(ctx context.Context, jobId string) ([]byte, error) {
-	location, err := joinURLPath(c.BaseURL, fmt.Sprintf("jobs/ingest/%s/unprocessedrecords", jobId))
+func (c *Connector) getJobResults(ctx context.Context, jobId string) (*http.Response, error) {
+	location, err := joinURLPath(c.BaseURL, fmt.Sprintf("jobs/ingest/%s/failedResults", jobId))
 	if err != nil {
 		return nil, err
 	}
 
-	return c.getCSV(ctx, location)
+	req, err := common.MakeJSONGetRequest(ctx, location, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create get request: %w", err)
+	}
+
+	return c.Client.Client.Do(req)
 }
 
-func (c *Connector) GetFailedResults(ctx context.Context, jobId string) ([]byte, error) {
-	location, err := joinURLPath(c.BaseURL, fmt.Sprintf("jobs/ingest/%s/failedResults", jobId))
+//nolint:funlen
+func (c *Connector) getPartialFailureDetails(ctx context.Context, jobInfo *GetJobInfoResult) (*JobResults, error) {
+	// Query SF to get partial failure details
+	res, err := c.getJobResults(ctx, jobInfo.Id)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get job results: %w", err)
+	}
+	defer res.Body.Close()
+	// if partially failed, we need to get the error message for each record
+	reader := csv.NewReader(res.Body)
+
+	failInfo := &FailInfo{
+		FailureType:         "Partial",
+		FailedUpdateRecords: make(map[string][]string),
+		FailedCreateRecords: make(map[string][]string),
+	}
+
+	var sfIdColIdx, sfErrorColIdx, externalIdColIdx int
+
+	rowIdx := 0
+
+	for {
+		record, err := reader.Read()
+		if err != nil {
+			// end of file
+			if errors.Is(err, io.EOF) {
+				break
+			}
+
+			return nil, err
+		}
+
+		// Get column index of sf__Id, sf__Error, and externalIdFieldName in header row
+		// SF API responses may not be consistent with the order of columns, so we need to get the index
+		if rowIdx == 0 {
+			indiceMap, err := getColumnIndice(record, []string{sfIdFieldName, sfErrorFieldName, jobInfo.ExternalIdFieldName})
+			if err != nil {
+				return nil, err
+			}
+
+			sfIdColIdx = indiceMap[sfIdFieldName]
+			sfErrorColIdx = indiceMap[sfErrorFieldName]
+			externalIdColIdx = indiceMap[jobInfo.ExternalIdFieldName]
+
```

Comment:
don't you need `rowIdx ++` before continue? otherwise this if loop will run again for the next row.